src/vpn-api/worker.ts

```diff
@@ -143,13 +162,77 @@ const worker = (instanceId: number) => {
 			),
 		)
 		.tap(() => openvpn.start())
+		.delay(1000)
+		.tap(() => {
+			return vpn.connect({ port: mgtPort, shellPrompt: '' }).then(() => {
+				vpn.connection.shellPrompt = '';
+				// monitor new client connections and map cid to uuid
+				vpn.on('log', data => {
+					if (data.includes('CLIENT:ESTABLISHED,')) {
+						const clientId = data.split(',')[1].trim();
+						const idMapper = (logData: string) => {
+							if (logData.includes('CLIENT:ENV,common_name=')) {
+								const now = new Date().getTime();
+								const uuid = logData.split('=')[1].trim();
+								// expire any previous mappings for this device
+								_.forOwn(clients, (client, key) => {
+									if (client.uuid === uuid) {
+										delete clients[key];
+									}
+								});
+								// register current mapping
+								clients[clientId] = {
+									uuid,
+									ts: now,
+									heartbeat: now,
+									tx: 0,
+									rx: 0,
+								};
+								logger.info(
+									`Parsed connect event for client_id=${clientId} uuid=${uuid}`,
+								);
+								vpn.removeListener('log', idMapper);
+							}
+						};
+						vpn.on('log', idMapper);
+					}
+				});
+				// process bytecount events to track realtime data usage
+				vpn.on('data', data => {
+					if (data.bytecount == null) {
+						return;
+					}
+					if (!data.bytecount[0].includes('CLI:')) {
+						return;
+					}
+					const clientId = data.bytecount[0].split(':')[1];
+					if (clients[clientId] == null) {
+						logger.error(`Unknown CID(${clientId}) from OpenVPN!`);
+						return;
+					}
+
+					const { uuid } = clients[clientId];
+					const ts = new Date().getTime();
+					const rxBytes = parseInt(data.bytecount[1], 10);
+					const txBytes = parseInt(data.bytecount[2], 10);
+					if (process != null && process.send != null) {
```

Comment:
Can `process` be null if we're running in nodejs?  What about `process.send`, when would that be null?

Reply:
Yeah I was pretty surprised at this one myself, but typescript complained that `process.send` might be null… The check on `process` is probably extraneous though.

Reply:
It makes sense, it can be undefined if you're the master or a child without an IPC channel: https://nodejs.org/api/process.html#process_process_send_message_sendhandle_options_callback

Reply:
There we go, that explains it. I've removed the check on `process` though, that wasn't necessary.